Mathlib/Combinatorics/Enumerative/Bell.lean

```diff
@@ -0,0 +1,217 @@
+  set p := n - 1
+  have hp : n = p + 1 := (succ_pred_eq_of_ne_zero hn).symm
+  simp only [hp, add_succ_sub_one]
+  calc
+    (m * (p + 1) + (p + 1)).choose (p + 1) * ((m * (p+1))! * (p+1)!)
+      = (m * (p + 1) + (p + 1)).choose (p + 1) * (m * (p+1))! * (p+1)! := by ring
+    _ = (m * (p+ 1) + (p + 1))! := by rw [add_choose_mul_factorial_mul_factorial]
+    _ = ((m * (p+ 1) + p) + 1)! := by ring_nf
+    _ = ((m * (p + 1) + p) + 1) * (m * (p + 1) + p)! := by rw [factorial_succ]
+    _ = (m * (p + 1) + p)! * ((p + 1) * (m + 1)) := by ring
+    _ = ((m * (p + 1) + p).choose p * (m * (p+1))! * (p)!) * ((p + 1) * (m + 1)) := by
+      rw [add_choose_mul_factorial_mul_factorial]
+    _ = (m * (p + 1) + p).choose p * (m * (p+1))! * (((p + 1) * (p)!) * (m + 1)) := by ring
+    _ = (m * (p + 1) + p).choose p * (m * (p+1))! * ((p + 1)! * (m + 1)) := by rw [factorial_succ]
+    _ = (m + 1) * (m * (p + 1) + p).choose p * ((m * (p + 1))! * (p + 1)!) := by ring
+
+theorem Nat.choose_mul_right (m) {n : ℕ} (hn : n ≠ 0) :
+    (m * n).choose n = m * (m * n - 1).choose (n - 1) := by
+  by_cases hm : m = 0
+  · simp only [hm, zero_mul, choose_eq_zero_iff]
+    exact Nat.pos_of_ne_zero hn
+  · set p := m - 1; have hp : m = p + 1 := (succ_pred_eq_of_ne_zero hm).symm
+    simp only [hp]
+    rw [add_mul, one_mul, choose_mul_add _ hn]
+
+namespace Multiset
+
+/-- Number of partitions of a set of cardinality `m.sum`
+whose parts have cardinalities given by `m` -/
+def bell (m : Multiset ℕ) : ℕ :=
+  Nat.multinomial m.toFinset (fun k ↦ k * m.count k) *
+    (m.toFinset.erase 0).prod
+      (fun k ↦ Finset.prod (Finset.range (m.count k)) fun j ↦ (j * k + k - 1).choose (k-1))
+
+@[simp]
+theorem bell_zero : bell 0 = 1 := by unfold bell; simp
+
+private theorem bell_mul_eq_lemma {x : ℕ} (hx : x ≠ 0) (c : ℕ) :
+    x ! ^ c * c ! * ∏ j ∈ Finset.range c, (j * x + x - 1).choose (x - 1) = (x * c)! := by
+  induction c with
+  | zero => simp
+  | succ c hrec =>
+    suffices  x ! ^ (c + 1) * (c + 1) ! = x ! * (c + 1) * (x ! ^ c * c !) by
+      rw [this]
+      rw [← mul_assoc]
+      simp only [Finset.range_succ, Finset.mem_range, lt_self_iff_false, not_false_eq_true,
+        Finset.prod_insert]
+      simp only [mul_assoc]
+      rw [mul_comm ((c * x + x - 1).choose (x-1))]
+      rw [← mul_assoc]
+      rw [mul_comm]
+      simp only [← mul_assoc]
+      rw [hrec]
+      have : (x * c)! * ((c * x + x - 1).choose (x - 1)) * x ! * (c + 1)
+        = ((c + 1) * ((c * x + x - 1).choose (x - 1))) * (x * c)! *  x ! := by
+        ring_nf
+      rw [this]
+      rw [← Nat.choose_mul_add c hx]
+      rw [mul_comm c x]
+      rw [Nat.add_choose_mul_factorial_mul_factorial (x * c) x]
+      rw [mul_add, mul_one]
+    rw [factorial_succ, pow_succ]
+    ring_nf
+
+theorem bell_mul_eq (m : Multiset ℕ) :
+    m.bell * (m.map (fun j ↦ j !)).prod *
+      Finset.prod (m.toFinset.erase 0) (fun j ↦ (m.count j)!) = m.sum ! := by
```

Comment:
Similarly here and below, can you use the `∏ _ ∈ _, _` notation?

Reply:
Done for the third factor (the `Finset.prod`), is there a notation for Multiset.prop?